api/src/utils/apply-snapshot.ts

```diff
@@ -20,41 +21,80 @@ export async function applySnapshot(
 	const current = options?.current ?? (await getSnapshot({ database, schema }));
 	const snapshotDiff = options?.diff ?? getSnapshotDiff(current, snapshot);
 
+	type CollectionDelta = {
+		collection: string;
+		diff: Diff<Collection | undefined>[];
+	};
+
+	const toBeCreateCollections = filter(
+		snapshotDiff.collections,
+		({ diff }) => diff[0].kind === 'N' && isNull((diff[0] as DiffNew<Collection>).rhs.meta?.group)
+	).map((item) => item as CollectionDelta);
+	const toBeDeleteCollections = filter(
+		snapshotDiff.collections,
+		({ diff }) => diff[0].kind === 'D' && isNull((diff[0] as DiffDeleted<Collection>).lhs.meta?.group)
+	).map((item) => item as CollectionDelta);
+
+	const getToBeCreateCollection = function (currentLevelCollections: string[]) {
+		return filter(snapshotDiff.collections, ({ diff }) => {
+			if ((diff[0] as DiffNew<Collection>).rhs) {
+				return includes(currentLevelCollections, (diff[0] as DiffNew<Collection>).rhs.meta?.group);
+			}
+		}).map((item) => item as CollectionDelta);
+	};
+	const getToBeDeleteCollection = function (currentLevelCollections: string[]) {
+		return filter(snapshotDiff.collections, ({ diff }) => {
+			if ((diff[0] as DiffDeleted<Collection>).lhs) {
+				return includes(currentLevelCollections, (diff[0] as DiffDeleted<Collection>).lhs.meta?.group);
+			}
+		}).map((item) => item as CollectionDelta);
+	};
+
 	await database.transaction(async (trx) => {
 		const collectionsService = new CollectionsService({ knex: trx, schema });
 
-		for (const { collection, diff } of snapshotDiff.collections) {
-			if (diff?.[0].kind === 'D') {
-				try {
-					await collectionsService.deleteOne(collection);
-				} catch (err) {
-					logger.error(`Failed to delete collection "${collection}"`);
-					throw err;
+		const createCollections = async function (collections: CollectionDelta[]) {
+			if (!isEmpty(collections)) {
+				for (const item of collections) {
```

Comment:
```suggestion
				for (const { collection, diff } of collections) {
```

More of a nitpick but let's keep the original destructuring to keep things similar as before 👍 (unless of course there's a naming collision that I'm missing here)